sdk/resource/config.go

```diff
@@ -0,0 +1,145 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/label"
+)
+
+// config contains configuration for Resource creation.
+type config struct {
+	// detectors that will be evaluated.
+	detectors []Detector
+
+	// telemetrySDK is used to specify non-default
+	// `telemetry.sdk.*` attributes.
+	telemetrySDK Detector
+
+	// HostResource is used to specify non-default `host.*`
+	// attributes.
+	host Detector
+
+	// FromEnv is used to specify non-default OTEL_RESOURCE_ATTRIBUTES
+	// attributes.
+	fromEnv Detector
+}
+
+// Option is the interface that applies a configuration option.
+type Option interface {
+	// Apply sets the Option value of a config.
+	Apply(*config)
+}
+
+// WithAttributes adds attributes to the configured Resource.
+func WithAttributes(attributes ...label.KeyValue) Option {
+	return WithDetectors(detectAttributes{attributes})
+}
+
+type detectAttributes struct {
+	attributes []label.KeyValue
+}
+
+func (d detectAttributes) Detect(context.Context) (*Resource, error) {
+	return New(d.attributes...), nil
+}
+
+// WithDetectors adds detectors to be evaluated for the configured resource.
+func WithDetectors(detectors ...Detector) Option {
+	return detectorsOption{detectors}
+}
+
+type detectorsOption struct {
+	detectors []Detector
+}
+
+func (o detectorsOption) Apply(cfg *config) {
+	cfg.detectors = append(cfg.detectors, o.detectors...)
+}
+
+// WithTelemetrySDK overrides the builtin `telemetry.sdk.*`
+// attributes.  Use nil to disable these attributes entirely.
+func WithTelemetrySDK(d Detector) Option {
+	return telemetrySDKOption{d}
+}
+
+type telemetrySDKOption struct {
+	Detector
+}
+
+func (o telemetrySDKOption) Apply(cfg *config) {
+	cfg.telemetrySDK = o.Detector
+}
+
+// WithHost overrides the builtin `host.*` attributes.  Use nil to
+// disable these attributes entirely.
+func WithHost(d Detector) Option {
+	return hostOption{d}
+}
+
+type hostOption struct {
+	Detector
+}
+
+func (o hostOption) Apply(cfg *config) {
+	cfg.host = o.Detector
+}
+
+// WithFromEnv overrides the builtin detector for
+// OTEL_RESOURCE_ATTRIBUTES.  Use nil to disable environment checking.
+func WithFromEnv(d Detector) Option {
+	return fromEnvOption{d}
+}
+
+type fromEnvOption struct {
+	Detector
+}
+
+func (o fromEnvOption) Apply(cfg *config) {
+	cfg.fromEnv = o.Detector
+}
+
+// WithoutBuiltin disables all the builtin detectors, including the
+// telemetry.sdk.*, host.*, and the environment detector.
+func WithoutBuiltin() Option {
+	return noBuiltinOption{}
+}
+
+type noBuiltinOption struct{}
+
+func (o noBuiltinOption) Apply(cfg *config) {
+	cfg.host = nil
+	cfg.telemetrySDK = nil
+	cfg.fromEnv = nil
+}
+
+// NewConfig returns a Resource combined from the provided attributes,
+// user-provided detectors and builtin detectors.
+func NewConfig(ctx context.Context, opts ...Option) (*Resource, error) {
```

Comment:
I don't think I was paying too close attention in my last review when I suggested we rename this from `Configure` to `NewConifg`. This doesn't return a `config`, to that probably not the best name. I'm guessing a better name would reflect the `Resource` being returned and the use of `Detectors` to create it. Possibly `NewFromDetectors` to compliment `New`.

I don't want the naming to block this PR. I'll open an issue to track the renaming.

Reply:
https://github.com/open-telemetry/opentelemetry-go/issues/1263

Reply:
OK I'm happy with `NewFromDetectors`, otherwise this PR is ready to merge.